### DISCUSSION TITLE
Build libraries to support arch packages.

### DISCUSSION THREAD

**Discussion by Maxython** [Maxython/arch-packages-for-termux#2] [General]
## Hello everyone. 
At the moment, a package is being developed to support arch packages, this is libfs (lib for support). The first tests (at a low level) have already been carried out and it showed itself well.  
## Test:
[Video](https://www.reddit.com/r/termux/comments/nt8jld/successful_installation_of_arch_packages_on/?utm_medium=android_app&utm_source=share)
Commands and list of libraries:  
```
$ ldd /bin/iw
        linux-vdso.so.1 (0x00000072a0f60000)
        libnl-genl-3.so.200 => /usr/lib/libnl-genl-3.so.200 (0x00000072a0ee6000)
        libnl-3.so.200 => /usr/lib/libnl-3.so.200 (0x00000072a0eb0000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00000072a0d3c000)
        /lib/ld-linux-aarch64.so.1 => /usr/lib/ld-linux-aarch64.so.1 (0x0000003000000000)
        libpthread.so.0 => /usr/lib/libpthread.so.0 (0x00000072a0d0c000)
```
```
$ ldd /bin/sed
        linux-vdso.so.1 (0x000000734d8c1000)
        libacl.so.1 => /usr/lib/libacl.so.1 (0x000000734d865000)
        libc.so.6 => /usr/lib/libc.so.6 (0x000000734d6f1000)
        /lib/ld-linux-aarch64.so.1 => /usr/lib/ld-linux-aarch64.so.1 (0x0000003000000000)
```
----
The following libraries are required for the next test:  
- `libdl-2.32.so`
- `libm-2.32.so`
- `libc.so`

These libraries are often used, but they break the termux system or command.  
- `libdl-2.32.so` - breaks the termux system.
- `libm-2.32.so` - breaks the pacman command.
- `libc.so` - this error was discussed [here](https://github.com/termux/termux-app/issues/2107).

Therefore, now the main goal is to create an analogue of these libraries.

**Comment by Maxython** [comment]
@Yisus7u7

**Comment by JesusChapman** [comment]
Are you doing cross-compilation?, I must remember that to which you can adapt those libraries to Termux, regardless of the values it is very difficult that they do not fence to break the termux system

**Comment by Maxython** [comment]
I understand.  But we must understand whether it is possible to make support for the arch packages.

**Reply by JesusChapman** [reply]
Well, in that case, it can be done but it will break termux, and that could affect proot..... 
Without proot you can't use termux-chroot

**Reply by JesusChapman** [reply]
To one that could make a simple distribution that runs under proot in termux, that does not interfere with the termux file system, something similar to proot-distro

**Reply by Maxython** [reply]
We need to fix the sources of these libraries.

**Comment by Maxython** [comment]
### Another part of the test.
The termux system packages have been changed to arch packages.

- `coreutils`
- `util-linux`

Bottom line: termux did not start because of the `login` command, but if you remove it, everything works (some commands do not work due to lack of libraries, but all basic (system) ones work without problems).

**Comment by Maxython** [comment]
### List of bugs for further discussion.

- `libc.so`:  
 ```
CANNOT LINK EXECUTABLE "-bash": "/data/data/com.termux/files/usr/lib/libc.so" has bad ELF magic: 2f2a2047
```

- `libdl-2.32.so`:
```
CANNOT LINK EXECUTABLE "-bash": library "libc.so.6" not found: needed by /data/data/com.termux/files/usr/lib/libdl-2.32.so in namespace (default)
```

- `libm-2.32.so` (running pacman):
```
CANNOT LINK EXECUTABLE "pacman": library "libc.so.6" not found: needed by /usr/lib/libm-2.32.so in namespace (default)
```

**Reply by Maxython** [reply]
Let's start with an error due to `libdl-2.32.so` and `libm-2.32.so`.  Reason: the library `libc.so.6` was not found.  Which is rather strange because the library is in place.  There can be two reasons for the error:
1. A crash has occurred inside these libraries.
2. A crash has occurred in the `libc.so.6` library.

It is also possible that these libraries are addressing the wrong coordinates, but this is vriyatli.

**Reply by Maxython** [reply]
By the way, here is the reason for the termux system not starting because of `libdl-2.32.so`:
```
~ $ ldd /bin/bash
        linux-vdso.so.1 (0x0000007b5e440000)
        libandroid-support.so => /data/data/com.termux/files/usr/lib/libandroid-support.so (0x0000007b5e37b000)
        libreadline.so.8 => /data/data/com.termux/files/usr/lib/libreadline.so.8 (0x0000007b5e32d000)
        libiconv.so => /data/data/com.termux/files/usr/lib/libiconv.so (0x0000007b5e21c000)
        libdl.so => not found
        libc.so => not found
        libc.so => not found
        libncursesw.so.6 => /data/data/com.termux/files/usr/lib/libncursesw.so.6 (0x0000007b5e1c4000)
        libc.so => not found
        libc.so => not found
        libc.so => not found
```
Bash takes this library, well, because of the error above, a bash error occurs.

**Reply by JesusChapman** [reply]
Oh, well

**Reply by Maxython** [reply]
The good news is that termux has been configured so that `libdl-2.32.so` and `libm-2.32.so` will see `libc.so.6`.  
To do this, had to add the following value to export:
```
export LD_LIBRARY_PATH="/data/data/com.termux/files/usr/lib"
```
But the bad part is that the bugs have changed and some because of the `libc.so` file.  
### Example:
```
~ $ git
WARNING: linker: Warning: "/usr/lib/libpthread-2.32.so" has unsupported flags DT_FLAGS_1=0x29 (ignoring unsupported flags)
CANNOT LINK EXECUTABLE "git": cannot find "libdl.so" from verneed[0] in DT_NEEDED list for "/apex/com.android.runtime/lib64/bionic/libc.so"
```

```
~ $ bash
CANNOT LINK EXECUTABLE "bash": cannot find "libdl.so" from verneed[1] in DT_NEEDED list for "/usr/bin/bash"
```

```
~/test4 $ pacman
CANNOT LINK EXECUTABLE "pacman": cannot find "libdl.so" from verneed[1] in DT_NEEDED list for "/usr/lib/libcrypto.so.1.1"
```

```
~ $ wget
wget: /data/data/com.termux/files/usr/lib/libdl.so: version `LIBC' not found (required by /data/data/com.termux/files/usr/lib/libc.so)
wget: /data/data/com.termux/files/usr/lib/libdl.so: version `LIBC_N' not found (required by /data/data/com.termux/files/usr/lib/libc.so)
```
--------
But thanks to "add libraries" some arch packages are started.  
### gcc (works but not completely.):
```
~/test4 $ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/lto-wrapper
Target: aarch64-unknown-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://github.com/archlinuxarm/PKGBUILDs/issues --enable-languages=c,c++,fortran,go,lto,objc,obj-c++,d --with-isl --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-install-libiberty --enable-linker-build-id --enable-lto --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-libunwind-exceptions --disable-multilib --disable-werror --host=aarch64-unknown-linux-gnu --build=aarch64-unknown-linux-gnu --with-arch=armv8-a --enable-fix-cortex-a53-835769 --enable-fix-cortex-a53-843419 gdc_include_dir=/usr/include/dlang/gdc
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 10.2.0 (GCC)
```
### make (it seems to work):
```
GNU Make 4.3
Built for aarch64-unknown-linux-gnu
Copyright (C) 1988-2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Comment by JesusChapman** [comment]
Anyway soon, maybe, I'll start with the method of building packages, and not using arch packages

**Comment by JesusChapman** [comment]
To a that everything you do would be easier if you do a distro that runs under proot...

**Comment by Maxython** [comment]
Ok, I understand you.

**Reply by JesusChapman** [reply]
Yes bro

**Comment by Maxython** [comment]
I configured libfs so that it doesn't break the termux system and that the arch commands work.  True, for the arch package to work, you need to run it with the `asl` command, then you need to make an automatic system.

[Video test with scripts in C and C++ (gcc)](https://www.reddit.com/r/termux/comments/o5n2k7/test_running_compiled_c_and_c_using_gcc_on_termux/?utm_medium=android_app&utm_source=share)

**Comment by Maxython** [comment]
## Second test for installing arch packages (medium level).
Installed packages:
- wget
- git
- nano
- file

### Outcomes:
- **wget**
    Works.  Installs files without errors.  There was no fix.
- **git**
    Works.  Installs files without errors.  There was a bug, fixed with the command `git config --global http.sslverify" false "`.
    Found out [here](https://github.com/scalingexcellence/scrapybook/issues/36#issuecomment-370126203).
- **nano**
    Works.  Creates and edits files without errors.  There was no fix.
- **file**
    Works.  Gives detailed information about files without errors.  There was no fix.

The video test will be only when some customized packages are created and when the installation of these libraries is added.

**Reply by Maxython** [reply]
Also installed `htop` for fun.  Bottom line: it works, but because access is denied, it can't read `/proc/stat`.

**Reply by JesusChapman** [reply]
That was to be expected, termux emulates those calls, arch packages will have problems with /proc

**Reply by Maxython** [reply]
Listen, when the desired packages are compiled, you too, please test the test.

**Reply by Maxython** [reply]
Installed `zsh` from pacman.  Bottom line: it works.

**Reply by JesusChapman** [reply]
Amazing you're in good work, because with that you can only use tigervnc and xfce4 would be even cooler

**Reply by Maxython** [reply]
They should also be checked but later.

**Comment by Maxython** [comment]
## Test compiling gcc on termux.
Trying to compile plain C on termux.
### Outcome:
The following error appears:
```
/data/data/com.termux/files/usr/bin/ld: /usr/lib/gcc/aarch64-unknown-linux-gnu/10.2.0/libgcc.a(lse-init.o): undefined reference to symbol '__getauxval@@GLIBC_2.17'
/data/data/com.termux/files/usr/bin/ld: /data/data/com.termux/files/usr/lib/libc.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```
The reason for the errors is not yet clear.

**Reply by Maxython** [reply]
The result is the same for C++.

**Reply by Maxython** [reply]
The main problem is that `gcc` cannot find `libc.so`, for some reason it looks in `/usr/lib` but it is in `/usr/lib/conlib`.

**Reply by Maxython** [reply]
For gcc to work, you had to place the `libc` type libraries in a shared folder.
